torch/csrc/autograd/variable.cpp

```diff
@@ -103,50 +104,48 @@ int64_t Variable::Impl::get_device_slow() const {
   return data_.get_device();
 }
 
-std::shared_ptr<Function> Variable::Impl::get_grad_accumulator() {
-  auto autograd_meta = get_autograd_meta();
-  if (autograd_meta->grad_fn_) {
+std::shared_ptr<Function> Variable::AutogradMeta::get_grad_accumulator(Variable self_var) {
+  if (grad_fn_) {
     throw std::logic_error(
         "get_grad_accumulator() should be only called on leaf Variables");
   }
-  if (!autograd_meta->requires_grad_) {
+  if (!requires_grad_) {
     return nullptr;
   }
 
-  std::lock_guard<std::mutex> lock(autograd_meta->mutex_);
+  std::lock_guard<std::mutex> lock(mutex_);
 
-  auto result = autograd_meta->grad_accumulator_.lock();
+  auto result = grad_accumulator_.lock();
   if (result)
     return result;
 
-  c10::raw::intrusive_ptr::incref(this);
-  auto intrusive_from_this = c10::intrusive_ptr<Variable::Impl>::reclaim(this);
+  c10::raw::intrusive_ptr::incref(self_var.unsafeGetTensorImpl());
+  auto intrusive_from_this = c10::intrusive_ptr<Variable::Impl>::reclaim(static_cast<Variable::Impl*>(self_var.unsafeGetTensorImpl()));
   result = std::make_shared<AccumulateGrad>(Variable(std::move(intrusive_from_this)));
-  autograd_meta->grad_accumulator_ = result;
+  grad_accumulator_ = result;
   return result;
 }
 
-void Variable::Impl::detach_() {
-  auto autograd_meta = get_autograd_meta();
-  if (autograd_meta->is_view_) {
+void Variable::AutogradMeta::detach_(Variable self_var) {
+  if (is_view_) {
     AT_ERROR("Can't detach views in-place. Use detach() instead");
   }
-  set_requires_grad(false);
-  autograd_meta->grad_fn_.reset();
-  autograd_meta->output_nr_ = 0;
+  set_requires_grad(false, self_var.unsafeGetTensorImpl());
+  grad_fn_.reset();
+  output_nr_ = 0;
 }
 
-void Variable::Impl::backward(
+void Variable::AutogradMeta::backward(
     c10::optional<Tensor> gradient,
     bool keep_graph,
-    bool create_graph) {
-  auto autograd_meta = get_autograd_meta();
+    bool create_graph,
+    Variable self_var) {
   std::vector<Edge> edges;
-  edges.emplace_back(autograd_meta->grad_fn_, autograd_meta->output_nr_);
+  edges.emplace_back(grad_fn_, output_nr_);
 
   std::vector<Variable> inputs;
   if (!gradient.has_value()) {
-    gradient = make_variable(at::ones_like(data_), /*requires_grad=*/false);
+    gradient = make_variable(at::ones_like(self_var.data()), /*requires_grad=*/false);
```

Comment:
On why I choose to pass in a `Variable self_var` to this function instead of passing in an `at::TensorImpl`: there are a couple places in the autograd functions where we need to call `at::ones_like()` and `at::TensorGeometry()` with the Tensor object, and calling them with TensorImpl wouldn't work.

Reply:
If you put this here, as opposed to a comment, no one will be able to find this useful information ever again!

Reply:
Going to add this as a comment in code when we decide on where to put the Variable::Impl functions (in AutogradMeta vs. in Variable) :)